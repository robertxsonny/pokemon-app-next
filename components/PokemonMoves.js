import { useState } from 'react';
import styled from '@emotion/styled';

import { lg } from '../styles/breakpoints';
import { TextButton } from '../styles/shared';

const MoveItemsOuter = styled.div({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'center'
})

const MoveItemsWrapper = styled.div({
  alignSelf: 'stretch',
  display: 'grid',
  gridTemplateColumns: '1fr 1fr',
  justifyContent: 'space-between',
  gap: 8,
  marginBottom: 16,
  transition: '0.3s all ease',
  [lg.min]: {
    gridTemplateColumns: '1fr 1fr 1fr',
  }
})

const MoveItem = styled.div({
  textTransform: 'capitalize',
  fontSize: 13,
  fontWeight: 500,
  color: 'dimgray',
  padding: '8px 12px',
  borderRadius: 6,
  border: '1px solid lightgrey'
})

const ShowMoreLessButton = styled(TextButton)({
  color: 'green',
  fontSize: 13
})

const PokemonMoves = ({ moves = [] }) => {
  const [showAll, setShowAll] = useState(false);

  const movesToDisplay = showAll ? moves : moves.slice(0, 12);

  return (
    <MoveItemsOuter>
      <MoveItemsWrapper>
        {movesToDisplay.map(({ move }) => <MoveItem key={move.name}>{move.name.replace(/-/g, ' ')}</MoveItem>)}
      </MoveItemsWrapper>
      {!showAll && movesToDisplay.length < moves.length && <ShowMoreLessButton onClick={() => setShowAll(true)}>Show all {moves.length} moves</ShowMoreLessButton>}
      {showAll && <ShowMoreLessButton onClick={() => setShowAll(false)}>Show only first 12 moves</ShowMoreLessButton>}
    </MoveItemsOuter>
  )
}

export default PokemonMoves;
