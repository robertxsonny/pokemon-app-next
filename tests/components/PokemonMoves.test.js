import { fireEvent, render, screen } from '@testing-library/react';
import PokemonMoves from '../../components/PokemonMoves';

const moves = [
  { move: { name: 'pound' } },
  { move: { name: 'karate-chop' } },
  { move: { name: 'double-slap' } },
  { move: { name: 'comet-punch' } },
  { move: { name: 'mega-punch' } },
  { move: { name: 'pay-day' } }
];

const moreMoves = [
  { move: { name: 'fire-punch' } },
  { move: { name: 'ice-punch' } },
  { move: { name: 'thunder-punch' } },
  { move: { name: 'scratch' } },
  { move: { name: 'vise-grip' } },
  { move: { name: 'guillotine' } },
  { move: { name: 'razor-wind' } },
  { move: { name: 'swords-dance' } },
  { move: { name: 'cut' } },
  { move: { name: 'gust' } },
  { move: { name: 'wing-attack' } },
  { move: { name: 'whirlwind' } },
  { move: { name: 'fly' } },
  { move: { name: 'bind' } }
]

describe('PokemonMoves component', () => {
  test('Should render all moves if less than 12 items', () => {
    const { container } = render(<PokemonMoves moves={moves} />);
    expect(container).toMatchSnapshot();
  })

  test('Should render 12 moves partially at first if more than 12 items', () => {
    const { container } = render(<PokemonMoves moves={[...moves, ...moreMoves]} />);
    expect(container).toMatchSnapshot();

    fireEvent.click(screen.getByRole('button', { name: 'Show all 20 moves' }))
    expect(container).toMatchSnapshot();

    fireEvent.click(screen.getByRole('button', { name: 'Show only first 12 moves' }))
    expect(container).toMatchSnapshot();
  })
});
