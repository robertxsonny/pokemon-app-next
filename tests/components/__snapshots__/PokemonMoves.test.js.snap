// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonMoves component Should render 12 moves partially at first if more than 12 items 1`] = `
<div>
  <div
    class="css-1oxeyy9-MoveItemsOuter e1mi334c3"
  >
    <div
      class="css-1nvhei2-MoveItemsWrapper e1mi334c2"
    >
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        pound
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        karate chop
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        double slap
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        comet punch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        mega punch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        pay day
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        fire punch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        ice punch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        thunder punch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        scratch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        vise grip
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        guillotine
      </div>
    </div>
    <button
      class="css-cu5g8q-TextButton-ShowMoreLessButton e1mi334c0"
    >
      Show all 
      20
       moves
    </button>
  </div>
</div>
`;

exports[`PokemonMoves component Should render 12 moves partially at first if more than 12 items 2`] = `
<div>
  <div
    class="css-1oxeyy9-MoveItemsOuter e1mi334c3"
  >
    <div
      class="css-1nvhei2-MoveItemsWrapper e1mi334c2"
    >
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        pound
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        karate chop
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        double slap
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        comet punch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        mega punch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        pay day
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        fire punch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        ice punch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        thunder punch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        scratch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        vise grip
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        guillotine
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        razor wind
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        swords dance
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        cut
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        gust
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        wing attack
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        whirlwind
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        fly
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        bind
      </div>
    </div>
    <button
      class="css-cu5g8q-TextButton-ShowMoreLessButton e1mi334c0"
    >
      Show only first 12 moves
    </button>
  </div>
</div>
`;

exports[`PokemonMoves component Should render 12 moves partially at first if more than 12 items 3`] = `
<div>
  <div
    class="css-1oxeyy9-MoveItemsOuter e1mi334c3"
  >
    <div
      class="css-1nvhei2-MoveItemsWrapper e1mi334c2"
    >
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        pound
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        karate chop
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        double slap
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        comet punch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        mega punch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        pay day
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        fire punch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        ice punch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        thunder punch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        scratch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        vise grip
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        guillotine
      </div>
    </div>
    <button
      class="css-cu5g8q-TextButton-ShowMoreLessButton e1mi334c0"
    >
      Show all 
      20
       moves
    </button>
  </div>
</div>
`;

exports[`PokemonMoves component Should render all moves if less than 12 items 1`] = `
<div>
  <div
    class="css-1oxeyy9-MoveItemsOuter e1mi334c3"
  >
    <div
      class="css-1nvhei2-MoveItemsWrapper e1mi334c2"
    >
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        pound
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        karate chop
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        double slap
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        comet punch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        mega punch
      </div>
      <div
        class="css-1ewczkj-MoveItem e1mi334c1"
      >
        pay day
      </div>
    </div>
  </div>
</div>
`;
